graphbolt/src/concurrent_id_hash_map.cc

```diff
@@ -199,6 +209,30 @@ inline void ConcurrentIdHashMap<IdType>::InsertAndSet(IdType id, IdType value) {
   hash_map_.data_ptr<IdType>()[getValueIndex(pos)] = value;
 }
 
+template <typename IdType>
+void ConcurrentIdHashMap<IdType>::InsertAndSetSmaller(IdType id, IdType value) {
+  IdType pos = (id & mask_), delta = 1;
+  IdType* hash_map_data = hash_map_.data_ptr<IdType>();
+  InsertState state = AttemptInsertAt(pos, id);
+  while (state == InsertState::OCCUPIED) {
+    Next(&pos, &delta);
+    state = AttemptInsertAt(pos, id);
+  }
+
+  if (state == InsertState::INSERTED) {
+    hash_map_data[getValueIndex(pos)] = value;
+    return;
+  }
+
+  IdType val_pos = getValueIndex(pos);
+  IdType old_val = hash_map_data[val_pos];
```

Comment:
```suggestion
  IdType old_val = std::numeric_limits<IdType>::max();
```

Reply:
We always need to access hash_map_data atomically, setting it to a max value ensures that it is initialized in the while loop below.

Reply:
Store operations should be atomic but why does read operation also need to be atomic?

Reply:
All operations need to be atomic if you modify a memory location atomically. Otherwise, there will be a data race.

Reply:
Yes sorry I got it wrong. They should be atomic. But I think `hash_map_data[pos]` is atomic itself?

Reply:
It is a plain pointer, it is not an atomic pointer. If we want the code to work on multiple platforms, we can't access the pointer directly like that.

Reply:
Ok I see